### ISSUE TITLE
Track down timeout-after-shutdown error

### ISSUE THREAD

**Issue by noahgibbs** [scarpe-team/scarpe#255]
There's an error, most easily recognised by "Returned two sets of results!" as an error message, where a test returns correct, non-error results and then also gets a timeout because it hasn't successfully stopped when it should.

Example: https://github.com/scarpe-team/scarpe/actions/runs/5370900698
Failure logs attached: [test failure logs.zip](https://github.com/scarpe-team/scarpe/files/11860721/test.failure.logs.zip)

There's something odd about that, which you can see by checking lib/scarpe/wv/control_interface.rb. There's a @do_shutdown variable that's set when shutdown happens (when the shutdown event is dispatched) that should block all further heartbeats and redraws. The timeout is triggered by a heartbeat.

Here's an excerpt from the test-failure event log:

~~~
[2515] DEBUG WV::ControlInterface: CTL event :redraw [] {}
[2530] WARN  WV::ControlInterface: die_after - timed out after 1.0
[2531] DEBUG WV::ControlInterface: Writing results file "/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/scarpe_test_results.json20230625-2415-rmoi83" to disk!
[2571] DEBUG WV::ControlInterface: CTL event :shutdown [] {}
[2574] ERROR WV::ControlInterface: Writing multi-result failure file to "/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/scarpe_test_results.json20230625-2415-rmoi83"!
[2574] DEBUG WV::ControlInterface: CTL event :heartbeat [] {}
~~~

So we're getting a heartbeat printed at 2574 milliseconds in, but a shutdown at 2571. There should be no heartbeat afterward...

Basically, several things look odd here. Why do we see a heartbeat after shutdown? <s>Is that a succeed-after-fail or vice-versa?</s> Was the heartbeat late? Time to track some of them down.

Note: it was a success-after-timeout. So maybe the timeout was just too short? I've adjusted them in the PR (#256).

The heartbeat wasn't so much late, as the only one. We *do* have logging that will tell that, since we now log every heartbeat with its time in milliseconds. But that doesn't tell us if it was late relative to the test timeout, so I'll log that too.

**Comment by noahgibbs** [comment]
I'm adding a check so that tests complain on the console if you're getting too near your timeouts. It won't be perfect. For starters, CI is a lot slower than most local dev laptops.

But if we get timeouts on some CI builds and warnings that particular tests are near timeout on others, that's probably good enough for now.

I think we had at least one test that was just too close to timing out all the time, and that's part of what's going on with flaky failures.

**Comment by noahgibbs** [comment]
The odd order in the event log may be that events are starting and then "mixing", given how close they are (total start-to-end span is around 40 milliseconds) in the event log.

But still, there are clearly two consecutive CTL event logs:

~~~
[2571] DEBUG WV::ControlInterface: CTL event :shutdown [] {}
[2574] ERROR WV::ControlInterface: Writing multi-result failure file to "/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/scarpe_test_results.json20230625-2415-rmoi83"!
[2574] DEBUG WV::ControlInterface: CTL event :heartbeat [] {}
~~~

Normally that shouldn't happen. This isn't multithreading where the Ruby code can be preempted and switch to a different event mid-way. Until the shutdown returns control to the Webview event loop, the heartbeat shouldn't start. The only obvious way to interleave handlers (https://github.com/scarpe-team/scarpe/wiki/ScarpeEvents.md#scarpe-event-ordering) is if the shutdown event somehow *triggers* the heartbeat event from some handler, which shouldn't ever happen.

It may still be safer to have the shutdown event kill the process before exiting somehow. Right now we don't do that, basically because it's really hard to tell the event system "do something after X event, but only after all the regular handlers have already finished." But that could be added. Nearly all event systems have something like it, basically for this same reason: sometimes you want to do something *after* an event is finished including all handlers. They also tend to have before-event handlers, whether to filter/cancel events or to do prep work ahead of them. I don't think we need before events yet, though.

**Comment by noahgibbs** [comment]
Oh hey... I'm returning early with the heartbeat, but I'm still logging it. So that's why a heartbeat can still be logged after a shutdown is logged. I'll clarify that log message, so in the shutdown case it'll have a "shutdown has already been called, not dispatching heartbeat"-type message.

**Comment by noahgibbs** [comment]
The other tests on CI don't seem to be near their timeout limits - the worst is test_destroy_from_display at 45% of current timeout threshold. But we should get prints at the end of the run if anything is using 50%+ of its timeout.

**Comment by noahgibbs** [comment]
Also worth mentioning: the empty_app test had a timeout of 1 second, which was probably too low on CI. I just cut out its manual override and changed its timeout to the default (3 seconds after the PR.)

Also, timing starts after Webview has booted up and started running display code, so it's not as short as it sounds.
